### PR TITLE
Add inline grid meridian/parallel labeling

### PR DESCRIPTION
**Pull request by ahuang11**
Work in progress after the newest merge; I know I haven't fully updated the tests! (Ignore the red s, I was using that as a reference)

image
image
image

### CODE REVIEW

**Comment by ahuang11** [comment]
Maybe shouldn't remove that _assert_can_draw_ticks; instead, add an additional keyword arg to ax.gridlines(inline=True)~

**Comment by ahuang11** [comment]
Noticed some projections might be a bit off too
Half of the temperatures in US seems to be wrong:
image

England is all dark red in one and light red in the other
image
image

File: lib/cartopy/mpl/gridliner.py [hunk 1]
```diff
@@ -95,8 +120,10 @@ class Gridliner(object):
     # maybe even a plain old mpl axes) and it will call the "_draw_gridliner"
     # method on draw. This will enable automatic gridline resolution
     # determination on zoom/pan.
-    def __init__(self, axes, crs, draw_labels=False, xlocator=None,
-                 ylocator=None, collection_kwargs=None):
+    def __init__(self, axes, crs, draw_labels=False,
```
Comment (dopplershift) [inline lib/cartopy/mpl/gridliner.py]: This is changing the `Gridlines` API. Thoughts, @pelson ?

File: lib/cartopy/mpl/gridliner.py [hunk 2]
```diff
@@ -156,6 +191,25 @@ def __init__(self, axes, crs, draw_labels=False, xlocator=None,
         #: Whether to draw labels on the right hand side of the map.
         self.ylabels_right = draw_labels
 
+        if not auto:
+            #: Whether to draw x labels inline
+            self.x_inline = x_inline
+            #: Whether to draw y labels inline
+            self.y_inline = y_inline
+        else:
+            if isinstance(self.axes.projection, _SUPPORTED_PROJS):
```
Comment (dopplershift) [inline lib/cartopy/mpl/gridliner.py]: I think this block could be replaced by:
```python
self.x_inline = not isinstance(self.axes.projection, _SUPPORTED_PROJS) and isinstance(self.axes.projection, _X_INLINE_PROJS)
self.y_inline = not isinstance(self.axes.projection, _SUPPORTED_PROJS) and isinstance(self.axes.projection, _Y_INLINE_PROJS)
```

File: lib/cartopy/mpl/gridliner.py [hunk 3]
```diff
@@ -362,22 +434,88 @@ def _draw_gridliner(self, nx=None, ny=None, background_patch=None):
         x_label_points = [x for x in x_ticks if x_lim[0] <= x <= x_lim[1]]
         y_label_points = [y for y in y_ticks if y_lim[0] <= y <= y_lim[1]]
 
+        proj = self.axes.projection
+        if self.x_inline or self.x_inline:
```
Comment (kaedonkers) [inline lib/cartopy/mpl/gridliner.py]: The `or` statement seems redundant here.

File: lib/cartopy/mpl/gridliner.py [hunk 4]
```diff
@@ -276,17 +342,23 @@ def _add_gridline_label(self, value, axis, upper_end):
...
         # coordinate with an appropriate offset.
-        label_transform = mtrans.blended_transform_factory(
-            x_transform=tr_x, y_transform=tr_y)
+        if other_value is None:
+            label_transform = mtrans.blended_transform_factory(
+                x_transform=tr_x, y_transform=tr_y)
+            alpha = 1
+        else:
+            label_transform = cartopy.crs.PlateCarree()
+            alpha = 0.7
```
Comment (QuLogic) [inline lib/cartopy/mpl/gridliner.py]: I'm not sure if we want to be changing transparency. Best would be if the line were cut under the label, like contour lines, but this is somewhat more difficult to do. But maybe using the contour machinery would help.

File: lib/cartopy/mpl/gridliner.py [hunk 5]
```diff
@@ -362,22 +434,88 @@ def _draw_gridliner(self, nx=None, ny=None, background_patch=None):
...
-            self._assert_can_draw_ticks()
             self.xformatter.set_locs(x_label_points)
             for x in x_label_points:
+                if self.x_inline:
+                    other_value = np.median(y_label_points)
+                    coordinate = (x, other_value)
+                    if coordinate in existing_coords:
+                        continue
+                    elif (isinstance(proj, cartopy.crs.TransverseMercator) and
+                          x in (-120, 120)):
```
Comment (QuLogic) [inline lib/cartopy/mpl/gridliner.py]: What are these hard-coded numbers for?

File: lib/cartopy/mpl/gridliner.py [hunk 6]
```diff
@@ -362,22 +434,88 @@ def _draw_gridliner(self, nx=None, ny=None, background_patch=None):
...
+                        continue
+                    else:
+                        existing_coords.append(coordinate)
+                        # remove overlaps
```
Comment (QuLogic) [inline lib/cartopy/mpl/gridliner.py]: On which projections?

File: lib/cartopy/mpl/gridliner.py [hunk 7]
```diff
@@ -362,22 +434,88 @@ def _draw_gridliner(self, nx=None, ny=None, background_patch=None):
...
                     self._add_gridline_label(x, axis='x', upper_end=True)
 
         if self.ylabels_left or self.ylabels_right:
-            self._assert_can_draw_ticks()
             self.yformatter.set_locs(y_label_points)
             for y in y_label_points:
+                if self.y_inline:
+                    if y == 0:  # (0, 0) already plotted from x
+                        continue
+                    if isinstance(proj, cartopy.crs.RotatedPole):
```
Comment (QuLogic) [inline lib/cartopy/mpl/gridliner.py]: Have you tested this with various choices of parameters to this projection?

File: lib/cartopy/mpl/gridliner.py [hunk 8]
```diff
@@ -362,22 +434,88 @@ def _draw_gridliner(self, nx=None, ny=None, background_patch=None):
...
+                        # so the labels aren't all the way to the left
+                        other_value = np.median(
+                            x_label_points[:int(len(x_label_points) / 2) - 2])
+                    elif isinstance(proj, cartopy.crs.RotatedPole):
```
Comment (QuLogic) [inline lib/cartopy/mpl/gridliner.py]: The same condition as above?

File: lib/cartopy/mpl/gridliner.py (hunk 2 above)
Comment (ahuang11) [reply lib/cartopy/mpl/gridliner.py]: I don't think it handles the else statement; also I think it's much more readable and intuitive for person reading code if I spread it out like this (personally I had a hard time understanding the one liners)

**Comment by kaedonkers** [comment]
Thank you for the contribution @ahuang11! Could you please also sign a CLA agreement, following the guidance in the [SciTools documentation](https://scitools.org.uk/organisation.html).

**Comment by corinnebosley** [comment]
Hi @ahuang11, this PR is really valuable and I'd like to move it forward if you're still in a position to work on it.

If you could fill out this [CLA form](https://scitools.org.uk/documents/cla_form.pdf) and return it to us at cla@metoffice.gov.uk then we can get this PR moving again, and I'd be very happy.

**Comment by ahuang11** [comment]
Hmm I screwed up the pull... Can anyone advise on how to rebase / remove all the commits from others?

File: lib/cartopy/mpl/gridliner.py (hunk 7 above)
Comment (ahuang11) [reply lib/cartopy/mpl/gridliner.py]: Done see notebook linked.

File: lib/cartopy/mpl/gridliner.py [hunk 9]
```diff
@@ -233,16 +288,24 @@ def _add_gridline_label(self, value, axis, upper_end):
         upper_end: bool
             If True, place at the maximum of the "other" coordinate (Axes
             coordinate == 1.0).  Else 'lower' end (Axes coord = 0.0).
+        other_value: float
+            The other coordinate value of this gridline used for inline
+            positioning. So if `value` is the x, `other_value` is the y.
```
[the remark this answers is not in the document]
Comment (ahuang11) [reply lib/cartopy/mpl/gridliner.py]: Done.

File: lib/cartopy/mpl/gridliner.py [hunk 10]
```diff
@@ -219,7 +273,8 @@ def _crs_transform(self):
             transform = transform._as_mpl_transform(self.axes)
         return transform
 
-    def _add_gridline_label(self, value, axis, upper_end):
+    def _add_gridline_label(self, value, axis, upper_end,
+                            other_value=None, clip_on=False):
```
[the remark this answers is not in the document]
Comment (ahuang11) [reply lib/cartopy/mpl/gridliner.py]: Done.

File: lib/cartopy/mpl/gridliner.py (hunk 1 above)
Comment (ahuang11) [reply lib/cartopy/mpl/gridliner.py]: Done.

**Comment by corinnebosley** [comment]
Hi @ahuang, thanks for filling out the CLA form, I will remove the `Blocked: CLA needed` label as soon as I get permissions, but in the meantime you have been added to the list of SciTools contributors.

It looks like you need to do another rebase here and also revisit some of the tests.

The first test failure is in `cartopy/tests/mpl/test_gridliner.py` `test_gridliner`.  This is one that you haven't changed, but it's expecvting an error where there now isn't one (because of your lovely work), so you will have to change the expectation of this test.

The next two test failures are triggered by the decorator `@ImageTesting` on your additional tests and are failing because the images don't match.  Admittedly I'm not sure why and haven't looked into it in great detail, but I thought you might want to personally, having more context.

The final failure is something of a conundrum to me at the moment.  It is occurring in a test that you haven't touched, but it tests gridlines, so this may be something worth looking into in case it is genuinely picking up an error or bug.

Good luck!

**Comment by ahuang11** [comment]
Not sure why the mpl v1.5 tests are still failing. Replicated the environment
```
Package              Version        Location
-------------------- -------------- -----------------------------------------------
asn1crypto           0.24.0
atomicwrites         1.2.1
attrs                18.2.0
Cartopy              0.16.1.dev337- /mnt/c/Users/Solactus/GOOGLE~1/Bash/cartopy/lib
certifi              2018.8.24
cffi                 1.11.5
chardet              3.0.4
cryptography         2.3.1
cryptography-vectors 2.3.1
cycler               0.10.0
Cython               0.23.1
filelock             3.0.10
idna                 2.7
lxml                 4.2.5
matplotlib           1.5.1
more-itertools       4.3.0
nose                 1.3.7
numpy                1.10.0
OWSLib               0.17.0
pathlib2             2.3.3
Pillow               3.2.0
pip                  10.0.1
pluggy               0.8.0
py                   1.7.0
pycparser            2.19
pyepsg               0.4.0
pyOpenSSL            18.0.0
pyparsing            2.2.1
pyproj               1.9.5.1
pyshp                2.0.0
PySocks              1.6.8
pytest               4.0.2
python-dateutil      2.7.3
pytz                 2018.5
requests             2.19.1
scipy                0.16.0
setuptools           40.2.0
Shapely              1.6.4.post1
six                  1.11.0
urllib3              1.23
wheel                0.31.1
```
and runs fine on my local machine
image

**Comment by corinnebosley** [comment]
@ahuang11 I've restarted the tests, looks like the failures could have been a server-side issue.  I'll keep an eye on them.

**Comment by stefraynaud** [comment]
@ahuang11 here are the conda envs in which I successfully recreated all the reference figures used for the gridliner tests:
```bash
packages="pillow pytest pytest-xdist filelock pep8 pyshp shapely six requests pyepsg owslib"
options="-q -y -c conda-forge/label/testing -c conda-forge -c defaults --override-channels"
conda create $options -n cartopy_py27       python=2.7 $packages cython=0.17 numpy=1.10.0 matplotlib=1.5.1 nose proj4=4.9.1 scipy=0.16.0 libgfortran=1 mock futures
conda create $options -n cartopy_py35       python=3.5 $packages cython=0.23.1 numpy=1.10.0 matplotlib=1.5.1 nose proj4=4.9.1 scipy=0.16.0 libgfortran=1
conda create $options -n cartopy_py2_latest python=2   $packages cython numpy matplotlib proj4 scipy mock futures pytest nose
conda create $options -n cartopy_py3_latest python=3.6 $packages cython numpy matplotlib-base proj4 pykdtree scipy fiona
```
These envs are strictly those of travis.

**Comment by corinnebosley** [comment]
@ahuang11 Right, I've been asking around the team and I now know what you have to do.

The two tests which are still failing are doing so because the newly-generated images are not matching up with the images which were previously generated.  This could be due to new features or small changes in dependencies.

You have to first visually make sure that the newly-generated image (although slightly different than the previous one) is still what you expect to be produced here.  To do that, you can find the names of the two image files you need to compare in the test output (like those on L.1159 and 1160 of test 2551.1: https://travis-ci.org/SciTools/cartopy/jobs/468516482#L1157).

Use these two filenames to view the images using eog or similar:
> eog file1 file2

If you are happy that the differences are minimal and acceptable then you can manually replace the expected result with your newly generated result.  Then re-submit the PR and we'll see how we get on.

**Comment by ahuang11** [comment]
Thanks for the commands @stefraynaud; that seemed to work. I might have been using conda-forge earlier so it wasn't working.

**Comment by corinnebosley** [comment]
Hi @ahuang11, I've merged #1117 now, so if you could rebase this when you get time then we'll try and get things moving again.

**Comment by ahuang11** [comment]
> Hi @ahuang11, I've merged #1117 now, so if you could rebase this when you get time then we'll try and get things moving again.

Okay halfway there~

File: lib/cartopy/mpl/geoaxes.py [hunk 11]
```diff
@@ -1276,8 +1276,10 @@ def imshow(self, img, *args, **kwargs):
 
         return result
 
-    def gridlines(self, crs=None, draw_labels=False, xlocs=None,
-                  ylocs=None, dms=False, **kwargs):
+    def gridlines(self, crs=None, draw_labels=False,
+                  xlocs=None, ylocs=None, dms=False,
+                  x_inline=None, y_inline=None,
```
Comment (stefraynaud) [inline lib/cartopy/mpl/geoaxes.py]: Maybe a keyword to choose the kind of labels would be appropriate : "inline", "outline", "both".

**Comment by stefraynaud** [comment]
That's really nice with the addition of inline labels!

File: lib/cartopy/mpl/geoaxes.py (hunk 11 above)
Comment (ahuang11) [reply lib/cartopy/mpl/geoaxes.py]: I think in most cases, "outline" would be preferred and "both" would be very cluttered in my opinion. auto_inline defaults to True which will decide if inline or outline (probably mostly outline), but if users want to override it, then they can just set x_inline=True and y_inline=True.

**Comment by corinnebosley** [comment]
Hi @ahuang11, I'm still looking into the travis errors that are occurring at build time here, but it's bothering me that the SciTools CLA-checker isn't recognising you any more, despite the fact that I remember CLAing you myself.

I think what's happened is that one of my more knowledgeable and capable colleagues has moved elsewhere in the business, but not before making a whole load of modifications to SciTools which has broken a lot of links in chains.

You should be able to re-add yourself to the CLA list by filling in this form: https://docs.google.com/forms/d/e/1FAIpQLSfd0tdE-DcJOXh8ej_7T93IizwJFYBFyRWYQOi2A8QRaKwykA/viewform

Please ping me once you've done that, I'll need to think about what to do next.

**Comment by ahuang11** [comment]
@corinnebosley Done.

**Comment by corinnebosley** [comment]
Thanks @ahuang11, unfortunately it looks like the form isn't working.  I'll continue to investigate both issues and tell you when I've found something.

**Comment by corinnebosley** [comment]
@ahuang11 My mistake, apparently the form is working and you are now on two whole lists (I'm not sure why we now have two or where they live, but you are on them both).

I will continue to look into the test fails.  I'm still pretty confused.  I'll keep you posted.

**Comment by xylar** [comment]
Any headway here?  This would be such a great capability to have and it's one of the few missing features keeping my group from switching from basemap to cartopy for all our plotting.

**Comment by corinnebosley** [comment]
@ahuang11 @xylar I'm so sorry for dragging this on; we're very short-staffed at the moment but I'm trying to squeeze this in where I can.

I am a bit concerned that this test failure has several causes, but it may be possible to get it passing by adding the new image to the cartopy image repository.  Or by running idiff and accepting the image differences.

Before we do that though (and by 'we' I think I mean 'you' (@ahuang11)), I need to really dig into this error to make sure this is the right action to take.  It's possible that the tests that pass might be supposed to fail and actually our tests just aren't working; in which case it would be wrong to just add an image to the repo.

I will post updates when I know more.

**Comment by xylar** [comment]
@corinnebosley, thanks so much for the update.  I know we're all volunteers here and I appreciate this work and the time you all put into it.

**Comment by ahuang11** [comment]
Run tests

**Comment by ahuang11** [comment]
Okay, I think this time, the test isn't because the images are different, but rather it's failing to build cartopy.

**Comment by corinnebosley** [comment]
@ahuang11 You're quite right, the cartopy build itself is now failing due to updates to proj4.  My colleague is looking into this and either she or I will put up a fix when we understand what cython's going to do with it a little better.

For now though, all your tests are now passing (thanks for that!) and I want to give this another review before going ahead.

**Comment by corinnebosley** [comment]
Hey @ahuang11!

I'm mostly very happy, but I'm slightly concerned by the images produced by your inline grid label machine for the USA.  What worries me is the difference in text angles for labels which look like they should be at the same angle, shown on the Geostationary and Orthographic projections.

I know it's even more work, but could you please add some tests to check that the text angle is being calculated correctly?

It principle, it doesn't really matter if some plots just look ugly when constructed in a certain way (i.e. with inline grid labels); you just wouldn't make your plot with that argument.  However, if there genuinely is something being calculated incorrectly that's a different matter and I shouldn't merge it yet.

A few basic tests for the numbers themselves as opposed to images should do it I reckon.

**Comment by ahuang11** [comment]
If I'm understanding correctly, you're saying that the 80W should be the same angle?
image

I have checked out the current master branch of Cartopy and was able to replicate it because those aren't actually "inline" grid labels for Geostationary and Orthographic projections for the USA--those are labels from https://github.com/SciTools/cartopy/pull/1117 
image

I have the default set to `auto_inline=True` which by default doesn't do an inline grid label for Orthographic / Geostationary (it has to be in https://github.com/SciTools/cartopy/pull/1089/files#diff-ce4cd585b732c7d15594d639a4054cfdR42 https://github.com/SciTools/cartopy/pull/1089/files#diff-ce4cd585b732c7d15594d639a4054cfdR49)

This is what it would look like if it's forced to be inline (the angle will be set to 0).
image

Unfortunately, I did not write the angling logic and I don't have much of a clue about it...

**Comment by corinnebosley** [comment]
@ahuang11 I see.  On that basis (the basis that I merged #1117 and was happy with the logic and testing), LGTM!

**Comment by corinnebosley** [comment]
Thanks @ahuang11, nice work!

**Comment by milenaveneziani** [comment]
Hi all, thanks for putting this in! Do you have a rough idea of when the next release will be, with this feature in? Thanks!

**Comment by dopplershift** [comment]
This will be in the next release, 0.18. I'm hoping start pushing on that in a week or so, but I'm hesitant to give a firm timeline on when the release will be out.

**Pull request merged**